addons/hr_timesheet/hr_timesheet.py

```diff
@@ -103,129 +77,118 @@ def on_change_unit_amount(self, cr, uid, id, prod_id, unit_amount, company_id, u
+                    raise exceptions.ValidationError("The employee " + emp[0].name + " is not associated to a valid Product. Please define one for him or select another user.")
+                elif not emp[0].product_id.uom_id.id:
+                    raise exceptions.ValidationError("The employee " + emp[0].name + " is not associated to a valid Product Unit of Measure. Please define one for him or select another user.")
+                elif not emp[0].product_id.categ_id.property_account_expense_categ.id:
+                    raise exceptions.ValidationError("The employee " + emp[0].name + " is not associated to a valid Financial Account. Please define one for him or select another user.")
+
+    @api.onchange('user_id')
+    def V8_on_change_user_id(self):
+        if self.is_timesheet:
+            new_values = self.on_change_user_id(self.user_id.id, self.is_timesheet)
+            self.journal_id = new_values["value"]['journal_id']
+            self.product_id = new_values["value"]['product_id']
+            self.product_uom_id = new_values["value"]['uom_id']
+            self.general_account_id = new_values["value"]['general_account_id']
+
+            # New API style
+            #     emp = self.env['hr.employee'].search([('user_id','=',self.user_id.id or self.env.uid)])
+            #     if not emp:
+            #         model, action_id = self.env['ir.model.data'].get_object_reference('hr', 'open_view_employee_list_my')
+            #         msg = _("Employee is not created for this user. Please create one from configuration panel.")
+            #         raise exceptions.RedirectWarning(msg, action_id, _('Go to the configuration panel'))
+            #     else:
+            #         self.journal_id = self._get_analytic_journal()
+            #         self.product_id = self._get_employee_product()
+            #         self.product_uom_id = self._get_employee_unit()
+            #         self.general_account_id = self._get_general_account()
+
+    def on_change_user_id(self, cr, uid, ids, user_id, is_timesheet=False, context=None):
+        if is_timesheet:
+            if not user_id:
+                return {}
+            else:
+                res = {"value": {
+                    'journal_id' : self._get_analytic_journal(cr,uid, user_id, context=context),
+                    'product_id' : self._get_employee_product(cr,uid, user_id, context=context),
+                    'uom_id' : self._get_employee_unit(cr,uid, user_id, context=context),
+                    'general_account_id' : self._get_general_account(cr,uid, user_id, context=context)
+                    }
+                }
+                return res
+
+    @api.onchange('date')
+    def on_change_date(self):
+        if self.is_timesheet:
+            if self._origin.date and (self._origin.date != self.date):
+                raise exceptions.Warning('Changing the date will let this entry appear in the timesheet of the new date.')
```

Comment:
I don't really understand the meaning of the warning, compared to the condition ^^ .